cmake/LySet.cmake

```diff
@@ -21,3 +21,82 @@ macro(ly_set name)
         set(${name} "${ARGN}" PARENT_SCOPE)
     endif()
 endmacro()
+
+
+#! ly_map_cache_variable_configs: Looks up the specified name of the variable to
+# see if it is a cache variable with a per configuration suffix first.
+# This is followed by checking if the name itself is a cache variable.
+# Finally if additional arguments are supplied pass the named arguments
+# Then the following additional arguments will be parsed
+# "VALUE_PREFIX", "VALUE_SUFFIX" "${CMAKE_CONFIGURATION_TYPES}""
+# that maps CMake configs to values, it will use that has a default value
+# This function sets the name of the variable at PARENT SCOPE with
+# the name ${output_variable_name} to a value suitable to be used in a generator
+# expression that will expand the values on a per config basis
+# or return a default value if the CONFIG doesn't have a per configuration cache
+# value or default value
+
+# \arg:output_variable_name name of the variable to set in the parent scope with
+#      the per config cache variable value
+# \arg:cache_variable_name name of the cache variable to look up as well as
+#      the prefix of per config cache variables to look up as well
+# \arg:PREFIX (optional) value to prepend to each config value when adding to the
+#      generator expression
+# \arg:SUFFIX (optional) value to append to each config value when adding to the
+#      generator expression
+# \arg:[ARGN] (optional) Argument which contains a list of $<CONFIG>;<default_value>
+#      arguments for each argument that should use a default value
+#
+# Ex. If the arguments to this function is as follows:
+#     output_variable_name = DefaultSearchMode
+#     cache_variable_name = "LY_ARCHIVE_FILE_SEARCH_MODE"
+#     And the optional 3rdArgument list is set to "profile;2;release;2"
+#     Also if the CMAKE_CONFIGURATION_TYPES are equal to "debug;profile;release"
+# Then this function will lookup to see if there is a cache variable that exists
+# with the names of
+# "${LY_ARCHIVE_FILE_SEARCH_MODE_debug}", "${LY_ARCHIVE_FILE_SEARCH_MODE_profile}"
+# "${LY_ARCHIVE_FILE_SEARCH_MODE_release}" "${LY_ARCHIVE_FILE_SEARCH_MODE}"
+# The value of "${LY_ARCHIVE_FILE_SEARCH_MODE}" will be used for configs, that doesn't have
+# a cache value.
+# The default per config values will be used if neither the
+# "${LY_ARCHIVE_FILE_SEARCH_MODE_$<CONFIG>}" or "${LY_ARCHIVE_FILE_SEARCH_MODE}" values are set
+function(ly_map_cache_variable_configs output_variable_name cache_variable_name)
```

Comment:
this is making the cache variable and how it applies too complicated IMO.
The search mode is now going to always default to 2 in release. If we want to have that fixed, I would have that in the code directly and state that this cache variable only applies to debug/profile. 
However, I was expecting this cache variable to apply to all configurations and just have a different value for release, so basically, I was expecting:
```
set(LY_ARCHIVE_FILE_SEARCH_MODE "$<$<CONFIG:release>:2,0>" CACHE STRING "Se...
```
If the user forces a mode to something else, then that is the mode that will be set, no matter what configuration is setting. 
Then the mapping is straightforward because its always defined:
```
set(LY_SEARCH_MODE_DEFINE LY_ARCHIVE_FILE_SEARCH_MODE_DEFAULT=${LY_ARCHIVE_FILE_SEARCH_MODE})
```

Reply:
Well the LY_ARCHIVE_FILE_SEARCH_MODE_DEFAULT isn't defined in debug/profile.
It is only defined in Release to 2.

In debug/profile it uses the `#else` block in ArchiveVars where it sets the file search priority to the zero-initialized value of the `FileSearchPriority` enum.

I can replicate that behavior using the following generator expression for `LY_SEARCH_MODE_DEFINE` though
```
set(LY_SEARCH_MODE_DEFINE $<$<NOT:$<STREQUAL:${LY_ARCHIVE_FILE_SEARCH_MODE},"">>,LY_ARCHIVE_FILE_SEARCH_MODE_DEFAULT=${LY_ARCHIVE_FILE_SEARCH_MODE}>)
```

Pretty much if the `LY_ARCHIVE_FILE_SEARCH_MODE` evaluates to an empty string, then the `LY_SEARCH_MODE_DEFINE` is evaluates to nothing

Reply:
I have addressed your feedback.